frame index is the frame at the current time, frame kick is the frame the guest
programmed. mfindex is the total of all frames since controller startup. Adjust
mfindex_kick (total frames programmed by the guest) to the correct period,
because frames wrap every 16364 (14 bit).

diff --git a/src/lib/qemu/hw/usb/hcd-xhci.c b/src/lib/qemu/hw/usb/hcd-xhci.c
index 9ce7ca7..6867eca 100644
--- a/src/lib/qemu/hw/usb/hcd-xhci.c
+++ b/src/lib/qemu/hw/usb/hcd-xhci.c
@@ -1692,6 +1692,12 @@ static void xhci_calc_intr_kick(XHCIState *xhci, XHCITransfer *xfer,
     xfer->mfindex_kick = MAX(asap, kick);
 }
 
+/* 0 - 16383 (14 Bit from descriptor) */
+#define FRAME_INDEX(frame) (frame & 0x3fff)
+#define FRAME_PERIOD(frame) (frame & ~0x3fff)
+#define FRAME_PERIOD_NEXT(frame) (FRAME_PERIOD(frame) + 0x4000)
+#define FRAME_PERIOD_LAST(frame) (FRAME_PERIOD(frame) - 0x4000)
+
 static void xhci_calc_iso_kick(XHCIState *xhci, XHCITransfer *xfer,
                                XHCIEPContext *epctx, uint64_t mfindex)
 {
@@ -1705,12 +1711,20 @@ static void xhci_calc_iso_kick(XHCIState *xhci, XHCITransfer *xfer,
             xfer->mfindex_kick = asap;
         }
     } else {
-        xfer->mfindex_kick = ((xfer->trbs[0].control >> TRB_TR_FRAMEID_SHIFT)
-                              & TRB_TR_FRAMEID_MASK) << 3;
-        xfer->mfindex_kick |= mfindex & ~0x3fff;
-        if (xfer->mfindex_kick + 0x100 < mfindex) {
-            xfer->mfindex_kick += 0x4000;
+        unsigned frame_kick  = ((xfer->trbs[0].control >> TRB_TR_FRAMEID_SHIFT)
+                                & TRB_TR_FRAMEID_MASK) << 3;
+        unsigned frame_index = FRAME_INDEX(mfindex);
+
+        /* frame index wrapped, kick is still in previous period */
+        if (frame_index < 1000 && frame_kick > 15000) {
+           xfer->mfindex_kick = frame_kick | FRAME_PERIOD_LAST(mfindex);
+        }
+        /* kick index wrapped, kick index is in next period */
+        else if (frame_kick < 1000 && frame_index > 15000) {
+           xfer->mfindex_kick = frame_kick | FRAME_PERIOD_NEXT(mfindex);
         }
+        else
+           xfer->mfindex_kick = frame_kick | FRAME_PERIOD(mfindex);
     }
 }
 
